verl/single_controller/base/decorator.py

```diff
@@ -319,6 +319,91 @@ def collect_dp_compute_data_proto(worker_group, output):
+    from verl.single_controller.base.worker import DistGlobalInfo, DistRankInfo
+    assert isinstance(worker_group, WorkerGroup)
+
+    assert mesh_name in worker_group._dist_global_info, f"expecting {mesh_name} in {worker_group._dist_global_info}"
+    assert mesh_name in worker_group._dist_workers_info, f"expecting {mesh_name} in {worker_group._dist_workers_info}"
+
+    dist_global_info: DistGlobalInfo = worker_group._dist_global_info[mesh_name]
+    dist_workers_info: list[DistRankInfo] = worker_group._dist_workers_info[mesh_name]
+    
+    all_args = []
+    for arg in args:
+        assert isinstance(arg, (Tuple, List)) and len(arg) == dist_global_info.dp_size
+        transformed_args = []
+        for i in range(worker_group.world_size):
+            local_dp_rank = dist_workers_info[i].dp_rank
+            transformed_args.append(arg[local_dp_rank])
+        all_args.append(transformed_args)
+    all_args = tuple(all_args)
+
+    all_kwargs = {}
+    for k, v in kwargs.items():
+        assert isinstance(v, (Tuple, List)) and len(v) == dist_global_info.dp_size
+        transformed_v = []
+        for i in range(worker_group.world_size):
+            local_dp_rank = dist_workers_info[i].dp_rank
+            transformed_v.append(v[local_dp_rank])
+        all_kwargs[k] = transformed_v
+    return all_args, all_kwargs
+
+
+def dispatch_nd_compute_dataproto(mesh_name, worker_group, *args, **kwargs):
+    from verl.single_controller.base.worker_group import WorkerGroup
+    from verl.single_controller.base.worker import DistGlobalInfo, DistRankInfo
+    dist_global_info: DistGlobalInfo = worker_group._dist_global_info[mesh_name]
+    splitted_args, splitted_kwargs = _split_args_kwargs_data_proto(dist_global_info.dp_size, *args, **kwargs)
+    return dispatch_nd_compute(mesh_name, worker_group, *splitted_args, **splitted_kwargs)
+
+
+def collect_nd_compute(mesh_name, worker_group, output):
+    from verl.single_controller.base.worker_group import WorkerGroup
+    from verl.single_controller.base.worker import DistGlobalInfo, DistRankInfo
+    assert isinstance(worker_group, WorkerGroup)
+    assert len(output) == worker_group.world_size
+
+    assert mesh_name in worker_group._dist_global_info, f"expecting {mesh_name} in {worker_group._dist_global_info}"
+    assert mesh_name in worker_group._dist_workers_info, f"expecting {mesh_name} in {worker_group._dist_workers_info}"
+
+    dist_global_info: DistGlobalInfo = worker_group._dist_global_info[mesh_name]
+    dist_workers_info: list[DistRankInfo] = worker_group._dist_workers_info[mesh_name]
+
+    output_in_dp = []
+    pp_size = dist_global_info.pp_size
+    for global_rank in range(worker_group.world_size):
+        local_rank_info = dist_workers_info[global_rank]
+        if local_rank_info.tp_rank == 0 and local_rank_info.pp_rank == pp_size - 1 and local_rank_info.cp_rank == 0 and local_rank_info.ep_rank == 0:
+            output_in_dp.append(output[global_rank])
```

Comment:
As an optimization, maybe we can set the result of workers other than tp=0,pp=pp_size-1,cp=0,ep=0 to None to reduce communication cost.

Reply:
Can this be done from the driver side? Or user have to be aware of this?

Reply:
It's better done in client side, we can set result to None if worker is not tp=0,pp=pp_size-1,cp=0,ep=0
https://github.com/volcengine/verl/blob/main/verl/single_controller/ray/base.py#L453